modules/reactor/src/meshgenerators/PeripheralTriangleMeshGenerator.C

```diff
@@ -0,0 +1,453 @@
+
+  for (unsigned int i = 0; i < _outer_circle_num_segments; i++)
+  {
+    // rotation angle
+    Real theta = i * d_theta;
+    // calculate (x, y) coords
+    Real x = _outer_circle_radius * std::cos(theta);
+    Real y = _outer_circle_radius * std::sin(theta);
+
+    // add to outer boundary list
+    outer_polyline.emplace_back(x, y);
+    // create Node and add to mesh
+    Node * node = mesh->add_point(Point(x, y, 0.0));
+    // add to association map
+    point_node_map[outer_polyline.back()] = node;
+  }
+
+  //
+  // Additional Steiner points; create additional circles of Steiner points between inner and outer
+  // boundaries.
+  //
+
+  for (std::vector<unsigned int>::size_type idx = 0; idx < _extra_circle_radii.size(); idx++)
+  {
+    // extract Steiner point circle parameters
+    Real num_segments = _extra_circle_num_segments[idx];
+    Real radius = _extra_circle_radii[idx];
+
+    // radial spacing
+    d_theta = 2.0 * M_PI / num_segments;
+
+    for (unsigned int i = 0; i < num_segments; i++)
+    {
+      // rotation angle
+      Real theta = i * d_theta;
+      // calculate (x, y) coords
+      Real x = radius * std::cos(theta);
+      Real y = radius * std::sin(theta);
+
+      // add to Steiner point list
+      steiner_points.emplace_back(x, y);
+      // create Node and add to mesh
+      Node * node = mesh->add_point(Point(x, y, 0.0));
+      // add to association map
+      point_node_map[steiner_points.back()] = node;
+    }
+  }
+
+  //
+  // Create P2T CDT and add primary (outer) polyline
+  // NOTE: polyline must be a simple polygon. The polyline's points constitute constrained edges. No
+  // repeat points!!!
+  //
+
+  std::vector<p2t::Point *> outer_polyline_ptrs(outer_polyline.size());
+  std::transform(outer_polyline.begin(),
+                 outer_polyline.end(),
+                 outer_polyline_ptrs.begin(),
+                 [](p2t::Point & p) { return &p; });
+  p2t::CDT cdt{outer_polyline_ptrs};
+
+  // Add inner boundary
+  std::vector<p2t::Point *> inner_polyline_ptrs(inner_polyline.size());
+  std::transform(inner_polyline.begin(),
+                 inner_polyline.end(),
+                 inner_polyline_ptrs.begin(),
+                 [](p2t::Point & p) { return &p; });
+  cdt.AddHole(inner_polyline_ptrs);
+
+  // Add Steiner points
+  for (auto & point : steiner_points)
+    cdt.AddPoint(&point);
+
+  // Triangulate!
+  cdt.Triangulate();
+
+  //
+  // Translate triangulated C2T points/triangles to libMesh nodes/Tri3s
+  //
+
+  // Extract triangle list from triangulation
+  std::vector<p2t::Triangle *> triangles = cdt.GetTriangles();
+
+  // process each P2T triangle into libMesh Tri3
+  for (p2t::Triangle * triangle : triangles)
+  {
+
+    // create Tri3 element
+    Tri3 * tri_elem = new Tri3();
+
+    // process each point on triangle
+    for (const int i : std::vector<int>{0, 1, 2})
+    {
+      // extract p2t::point from triangle
+      p2t::Point * point = triangle->GetPoint(i);
```

Comment:
@aaronoaks -

Is poly2tri generating the ordering we expect here?  In my own poly2tri code I tried doing the same, set libMesh node i to p2t point i ... and it turns out I'm getting flipped elements, with clockwise instead of counter-clockwise node ordering.

It was actually quite tricky to run across that - libMesh generally assumes that your 2D elements might live in 3D space, makes sure their volume() isn't just half the cross product z component, likewise for JxW calculations, etc ... but even if it usually works fine I feel like this is going to bite us in some unexpected way (perhaps when we extrude the resulting mesh?) in the future.

Reply:
Actually, this might be a false alarm.  I am seeing a flipped element, but it may be from my own triangulation refinement code, not from poly2tri.  I'm adding some libMesh unit tests to make sure of that, though, and I still wouldn't mind independent confirmation of it.